pep-0590.rst

```diff
@@ -48,145 +48,212 @@ The function pointer type
-The new flag specifies that the function has the type ``PyObject *(*call) (PyObject *self, PyObject *const *args, Py_ssize_t nargs, PyObject *kwname)``
+A new constant ``METH_VECTORCALL`` is added for specifying ``PyMethodDef`` structs.
+It means that the C function has the signature ``vectorcallfunc``.
+This should be the preferred flag for new functions, as this avoids a wrapper function.
 
-Internal CPython changes
-========================
+**NOTE**: the numerical value of ``METH_VECTORCALL`` is unspecified
+and it may have more than 1 bit set.
+It must not combined with any of the existing flags
+``METH_VARARGS``, ``METH_FASTCALL``, ``METH_NOARGS``, ``METH_O`` or ``METH_KEYWORDS``.
+
+Subclassing
+-----------
 
-In order to conform to the specification, the only changes required are:
+Extension types inherit the type flag ``Py_TPFLAGS_HAVE_VECTORCALL``
+and the value ``tp_vectorcall_offset`` from the base class,
+provided that they implement ``tp_call`` the same way as the base class.
+Additionally, the flag ``Py_TPFLAGS_FUNCTION_DESCRIPTOR``
+is inherited if ``tp_descr_get`` and ``tp_descr_set`` are implemented the
+same way as the base class.
 
-* To use the new calling convention in the interpreter.
-* An implementation of the ``PyObject_VectorCall`` function.
-* An implementation of the ``PyCall_MakeVectorCall`` and ``PyCall_MakeTpCall`` convenience functions.
+Heap types never inherit the vectorcall protocol because
+that would not be safe (heap types can be changed dynamically).
+This restriction may be lifted in the future, but that would require
+special-casing ``__call__`` in ``type.__setattribute__``.
 
-To gain the promised performance advantage, the following classes will need to implement the new calling convention:
-* Python functions
-* Builtin functions and methods
-* Bound methods
-* Method descriptors
-* A few of the most commonly used classes, probably ``range``, ``list``, ``str``, and ``type``.
 
-Changes to existing C structs
------------------------------
+Internal CPython changes
+========================
+
+Changes to existing classes
+---------------------------
 
-The ``function``, ``builtin_function_or_method``, ``method_descriptor`` and ``method`` classes will have their corresponding structs changed to
-include a ``vectorcall`` pointer.
+The ``function``, ``builtin_function_or_method``, ``method_descriptor``, ``method``, ``wrapper_descriptor``, ``method-wrapper``
+classes will use the vectorcall protocol
+(not all of these will be changed in the initial implementation).
 
-Third-party built-in classes using the new extended call interface
-------------------------------------------------------------------
+For ``builtin_function_or_method`` and ``method_descriptor``
+(which use the ``PyMethodDef`` data structure),
+a specific vectorcall wrapper will be implemented for every existing calling convention.
```

Comment:
We could do this, but it might not be worth it. We shouldn't specify it in the PEP.

Reply:
I think it's an interesting idea, so I would keep it in the PEP. But I'll just say that it's something which "could" be done, not "will" be done.